feat(datepicker): allow outsideDays for any number of displayMonths

alexluecke opened the pull request:
# Motivation

I would expect the outside days to be present on any value of `displayMonths`. Foremost, the `outsideDays` for multiple months can be explicitly set to hidden to achieve the hidden outside days, so it does not make sense to inhibit potentially useful functionality. In addition, day display can be controlled through the `dayTemplate` input.

# Issues:

This is obviously a breaking change. However, since the project is still in alpha I would prefer nipping this one in the bud now to allow for more customizeable calendars in the future.

----

Before submitting a pull request, please make sure you have at least performed the following:

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated any applicable tests.
 - [x] added/updated any applicable API documentation.
 - [x] added/updated any applicable demos.

maxokorokov commented:
Hey, @alexluecke. Good suggestion, it was just simpler to implement this way in the beginning. I have some remarks though:

1. We have 3 values for `outsideDays` and we have decide on behaviour for multiple months:

- `hidden` → works now
- `collapsed` → would work the same as for 1 month I suppose
- `visible` → some days would be displayed twice in the view - selection have to be fixed, as now both of them would be highlighted. Plus what will happen when you click on them?

2. I think I personally would prefer to have `visible` as default for 1 month view and `hidden` for 1+ month view

3. Please implement corresponding unit tests, especially for the `visible` case, as we won't be able to merge the PR without them.

alexluecke commented:
So I've implemented the changes and added tests. For now I think it might be acceptable to have both days selected in the multi-calendar view.

Initially, I thought it might be a good idea to only mark days as selected if they aren't disabled, which would allow the user to disable the multi-date selection issue. However, when this is done the outside days won't be selectable and show up as greyed out days, so I don't think much is gained from just setting the `outsideDays` to `hidden` (as was the default). Being the case, the multi-selection would be my preferred path forward for now.

miller-time commented on src/datepicker/datepicker.ts:
```diff
@@ -84,7 +84,7 @@ export interface NgbDatepickerNavigateEvent {
             [showWeekNumbers]="showWeekNumbers"
             [disabled]="disabled"
-            [outsideDays]="displayMonths === 1 ? outsideDays : 'hidden'"
+            [outsideDays]="outsideDays || displayMonths === 1 ? 'visible' : 'hidden'"
```
I think with operator precedence you need some parens here:

```
[outsideDays]="outsideDays || (displayMonths === 1 ? 'visible' : 'hidden')"
```

Otherwise you'll be modifying the value of the month view's `outsideDays` to always be "visible" any time it is explicitly set on the outer datepicker.

alexluecke replied:
If this is true, then I am probably missing a test as well. Let me follow up.

alexluecke replied:
Good catch. Added test and fixed up the code.

wesleycho commented on src/datepicker/datepicker.spec.ts:
```diff
@@ -259,20 +259,31 @@ describe('ngb-datepicker', () => {
+  it('should set outside days to "hidden" if not provided when there are multiple months displayed', () => {
+    const fixture = createTestComponent(`<ngb-datepicker [displayMonths]="2"></ngb-datepicker>`);
+    let months = fixture.debugElement.queryAll(By.directive(NgbDatepickerMonthView));
     expect(months[0].componentInstance.outsideDays).toBe('hidden');
```
Just a minor nitpick if you want to make the adjustment, you could do
```ts
let [month] = fixture.debugElement.queryAll(By.directive(NgbDatepickerMonthView));
expect(month.componentInstance.outsideDays).toBe('hidden');
```
or
```ts
let [{componentInstance: {outsideDays}}] = fixture.debugElement.queryAll(By.directive(NgbDatepickerMonthView));
expect(outsideDays).toBe('hidden');
```
Whatever is preferred is I guess whatever makes it more readable, just food for thought.

alexluecke replied:
I think the first version is much cleaner, thanks for the suggestion. I will make the changes.

maxokorokov commented:
I'm sorry, I can't agree to merge as is, because this particular case doesn't look good to me at all:

<img width="485" alt="screen shot 2017-04-13 at 18 31 22" src="https://cloud.githubusercontent.com/assets/8074436/25014430/febe15c0-2077-11e7-8aff-e5cc8f860746.png">

With `[outsideDates]='visible'` and multiple months you'll get currently selected day displayed in double

Maybe in this case (multiple months, outsideDays == visible) we should simply 
- collapse all the days in between August and September
- leave July and October days visible

?

alexluecke commented:
@maxokorokov 

Could you clarify your asks? By "collapse" do you mean give the "collapse" view for outsideDays (instead of purely "visible")? For the use case, I just need to have all of the days for a given month (see: https://en.wikipedia.org/wiki/Broadcast_calendar for a calendar where this is applicable).

maxokorokov commented:
Yeah, maybe it wasn't clear, sorry. 

My suggestion is:
* red ones should be always forced to `collapsed` in case of multiple months
* yellow ones should depend on the actual `[outsideDays]` value

<img width="485" alt="febe15c0-2077-11e7-8aff-e5cc8f860746" src="https://user-images.githubusercontent.com/8074436/27035696-8f49ab70-4f82-11e7-8f2d-95bd1e9fb7f8.png">

pkozlowski-opensource commented:
@alexluecke we haven't heard back from you for a while so I wonder if we should close this one as stale? @maxokorokov WDYT?

alexluecke commented:
Yeah, I got a working solution that I was really not happy with. I will close ticket and open a new PR if I ever come up with a better solution.
